zebra-rpc/src/methods/get_block_template_rpcs/types/get_block_template.rs

```diff
@@ -238,27 +243,38 @@ impl GetBlockTemplate {
 
         // Convert transactions into TransactionTemplates
         let mut mempool_txs_with_templates: Vec<(
+            InBlockTxDependenciesDepth,
             TransactionTemplate<amount::NonNegative>,
             VerifiedUnminedTx,
         )> = mempool_txs
             .into_iter()
-            .map(|tx| ((&tx).into(), tx))
+            .map(|(min_tx_index, tx)| (min_tx_index, (&tx).into(), tx))
             .collect();
 
         // Transaction selection returns transactions in an arbitrary order,
         // but Zebra's snapshot tests expect the same order every time.
+        //
+        // # Correctness
+        //
+        // Transactions that spend outputs created in the same block must appear
+        // after the transactions that create those outputs.
         if like_zcashd {
             // Sort in serialized data order, excluding the length byte.
             // `zcashd` sometimes seems to do this, but other times the order is arbitrary.
-            mempool_txs_with_templates.sort_by_key(|(tx_template, _tx)| tx_template.data.clone());
+            mempool_txs_with_templates.sort_by_key(|(min_tx_index, tx_template, _tx)| {
+                (*min_tx_index, tx_template.data.clone())
+            });
         } else {
             // Sort by hash, this is faster.
-            mempool_txs_with_templates
-                .sort_by_key(|(tx_template, _tx)| tx_template.hash.bytes_in_display_order());
+            mempool_txs_with_templates.sort_by_key(|(min_tx_index, tx_template, _tx)| {
```

Comment:
Do we need to sort at all?

Reply:
Probably not, I think it was there to make comparisons with the zcashd getblocktemplate method easier. I wouldn't mind removing it.

Reply:
I'm in favor of removing it as well then.

Reply:
I've limited it to sorting only in test compilations (it's still needed for the snapshot tests).